engine/access/state_stream/handler.go

```diff
@@ -14,30 +18,107 @@ type Handler struct {
 	blockID, err := convert.BlockID(request.GetBlockId())
 	if err != nil {
-		return nil, err
+		return nil, status.Errorf(codes.InvalidArgument, "could not convert block ID: %v", err)
 	}
 
 	execData, err := h.api.GetExecutionDataByBlockID(ctx, blockID)
 	if err != nil {
-		return nil, err
+		return nil, rpc.ConvertError(err, "could no get execution data", codes.Internal)
+	}
+
+	message, err := convert.BlockExecutionDataToMessage(execData)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "could not convert execution data to entity: %v", err)
+	}
+
+	return &access.GetExecutionDataByBlockIDResponse{BlockExecutionData: message}, nil
+}
+
+func (h *Handler) SubscribeExecutionData(request *access.SubscribeExecutionDataRequest, stream access.ExecutionDataAPI_SubscribeExecutionDataServer) error {
+	startBlockID := flow.ZeroID
+	if request.GetStartBlockId() != nil {
+		blockID, err := convert.BlockID(request.GetStartBlockId())
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "could not convert start block ID: %v", err)
+		}
+		startBlockID = blockID
+	}
+
+	sub := h.api.SubscribeExecutionData(stream.Context(), startBlockID, request.GetStartBlockHeight())
+
+	for {
+		v, ok := <-sub.Channel()
+		if !ok {
+			return rpc.ConvertError(sub.Err(), "stream encountered an error", codes.Internal)
+		}
+
+		resp, ok := v.(*ExecutionDataResponse)
+		if !ok {
+			return status.Errorf(codes.Internal, "unexpected response type: %T", v)
+		}
+
+		execData, err := convert.BlockExecutionDataToMessage(resp.ExecutionData)
+		if err != nil {
+			return status.Errorf(codes.Internal, "could not convert execution data to entity: %v", err)
+		}
+
+		err = stream.Send(&executiondata.SubscribeExecutionDataResponse{
+			BlockHeight:        resp.Height,
+			BlockExecutionData: execData,
+		})
+		if err != nil {
+			return rpc.ConvertError(err, "could not send response", codes.Internal)
+		}
+	}
+}
+
+func (h *Handler) SubscribeEvents(request *access.SubscribeEventsRequest, stream access.ExecutionDataAPI_SubscribeEventsServer) error {
+	startBlockID := flow.ZeroID
+	if request.GetStartBlockId() != nil {
+		blockID, err := convert.BlockID(request.GetStartBlockId())
+		if err != nil {
+			return status.Errorf(codes.InvalidArgument, "could not convert start block ID: %v", err)
+		}
+		startBlockID = blockID
 	}
 
-	return &access.GetExecutionDataByBlockIDResponse{BlockExecutionData: execData}, nil
+	filter := EventFilter{}
+	if request.GetFilter() != nil {
+		reqFilter := request.GetFilter()
+		filter = NewEventFilter(reqFilter.GetEventType(), reqFilter.GetAddress(), reqFilter.GetContract())
+	}
+
+	sub := h.api.SubscribeEvents(stream.Context(), startBlockID, request.GetStartBlockHeight(), filter)
+
+	for {
+		v, ok := <-sub.Channel()
+		if !ok {
+			return rpc.ConvertError(sub.Err(), "stream encountered an error", codes.Internal)
```

Comment:
IIUC this would also be called on sub.Close()

Reply:
yea, if `sub.Err()` is nil, `rpc.ConvertError()` will return nil. I'll add an explicit check since that's not obvious